three-cards-divination/src/main/java/com/cyber/mysticism/tarot/service/ThreeCardsDivinationService.java

```diff
@@ -36,8 +42,26 @@ private static Collector<Card, Object, Stream<Card>> toShuffledStream() {
 
     /**
      * @return a map that contains the reading in the form past, present and future
+     * for a specific user, returning only non-duplicate readings
      */
-    public Map<String, Card> getReading() throws DivinationException {
+    public Map<String, Card> getReadingForUser(String username, String email) throws DivinationException, UserNotFoundException {
+        Optional<TarotUser> retrievedUser = userRepository.findById(username);
+        Optional<TarotUser> retrievedUserByEmail = userRepository.findByEmail(email);
+        if (retrievedUser.isEmpty() || retrievedUserByEmail.isEmpty() || retrievedUser.get() != retrievedUserByEmail.get()) {
+            logger.info("event=reading_for_user_failed, status=error, reason=invalid_user, type={}, user={}",
+                    THREE_CARDS_DIVINATION, username);
```

Comment:
**Suggestion:** The `getReadingForUser` method in `ThreeCardsDivinationService` checks if `retrievedUser` and `retrievedUserByEmail` are empty and if they are not the same. However, the `findById` and `findByEmail` methods both return an `Optional<TarotUser>`, so comparing them with `!=` might not give the expected result. Consider comparing the `TarotUser` objects inside the `Optional` instead. [bug]
```suggestion
    public Map<String, Card> getReadingForUser(String username, String email) throws DivinationException, UserNotFoundException {
        Optional<TarotUser> retrievedUser = userRepository.findById(username);
        Optional<TarotUser> retrievedUserByEmail = userRepository.findByEmail(email);
        if (retrievedUser.isEmpty() || retrievedUserByEmail.isEmpty() || !retrievedUser.get().equals(retrievedUserByEmail.get())) {
            logger.info("event=reading_for_user_failed, status=error, reason=invalid_user, type={}, user={}",
                    THREE_CARDS_DIVINATION, username);
            throw new UserNotFoundException("Cannot perform reading, user not found");
        }

        Map<String, Card> threeCardsDivinationReading = getReading();

        return getUniqueReadingBasedOnHashCode(retrievedUser.get(), threeCardsDivinationReading);
    }
```